build-tools/packages/version-tools/test/semver.test.ts

```diff
@@ -0,0 +1,347 @@
+            const expected = `minor`;
+            const result = detectConstraintType(input);
+            assert.strictEqual(result, expected);
+        });
+    });
+
+    describe("internal version scheme ranges", () => {
+        it("bump patch", () => {
+            const input = `>=2.0.0-internal.1.0.0 <2.0.0-internal.1.1.0`;
+            const expected = `>=2.0.0-internal.1.0.1 <2.0.0-internal.1.1.0`;
+            const result = incRange(input, "patch");
+            assert.strictEqual(result, expected);
+        });
+
+        it("bump minor", () => {
+            const input = `>=2.0.0-internal.1.0.1 <2.0.0-internal.2.0.0`;
+            const expected = `>=2.0.0-internal.1.1.0 <2.0.0-internal.2.0.0`;
+            const result = incRange(input, "minor");
+            assert.strictEqual(result, expected);
+        });
+
+        it("bump minor with patch constraint", () => {
+            const input = `>=2.0.0-internal.1.0.1 <2.0.0-internal.1.1.0`;
+            const expected = `>=2.0.0-internal.1.1.0 <2.0.0-internal.1.2.0`;
+            const result = incRange(input, "minor");
+            assert.strictEqual(result, expected);
+        });
+
+        it("bump minor with minor constraint", () => {
+            const input = `>=2.0.0-internal.1.0.1 <2.0.0-internal.2.0.0`;
+            const expected = `>=2.0.0-internal.1.1.0 <2.0.0-internal.2.0.0`;
+            const result = incRange(input, "minor");
+            assert.strictEqual(result, expected);
+        });
+
+        it("bump major with patch constraint", () => {
+            const input = `>=2.0.0-internal.1.0.1 <2.0.0-internal.1.1.0`;
+            const expected = `>=2.0.0-internal.2.0.0 <2.0.0-internal.2.1.0`;
+            const result = incRange(input, "major");
+            assert.strictEqual(result, expected);
+        });
+
+        it("bump major with minor constraint", () => {
+            const input = `>=2.0.0-internal.1.0.1 <2.0.0-internal.2.0.0`;
+            const expected = `>=2.0.0-internal.2.0.0 <2.0.0-internal.3.0.0`;
+            const result = incRange(input, "major");
+            assert.strictEqual(result, expected);
+        });
+    });
+
+    describe("virtualPatch version scheme ranges", () => {
+        describe("precise version", () => {
+            it("bump patch", () => {
+                const input = `0.1029.1000`;
+                const expected = `0.1029.1001`;
+                const result = incRange(input, "patch");
+                assert.strictEqual(result, expected);
+            });
+
+            it("bump minor", () => {
+                const input = `0.59.1001`;
+                const expected = `0.59.2000`;
+                const result = incRange(input, "minor");
+                assert.strictEqual(result, expected);
+            });
+
+            it("bump major", () => {
+                const input = `0.59.1001`;
+                const expected = `0.60.1000`;
```

Comment:
Ideally the expected should be 0.60.1000, but according to our versioning 0.59.1001's major bump is 1.0.0. should this be modified?

Reply:
Good question. I think in this case this is correct, because this code is not supposed to know about our versions, rather it knows about "version schemes". And in this case, the detected version scheme is "semver", so it's expected to follow the regular semver semantics. I don't want to duplicate the logic in version-utils, but once you merge your changes I might look at ways to integrate it. I am trying to keep this code pretty unaware of the specifics of how our versions have incremented over time.

Reply:
makes sense. merged my PR in :)